WordPress/src/test/java/org/wordpress/android/viewmodel/activitylog/ActivityLogViewModelTest.kt

```diff
@@ -662,30 +568,568 @@ class ActivityLogViewModelTest {
+        canLoadMore: Boolean = false,
+        withFooter: Boolean = false
+    ): List<ActivityLogListItem> {
+        val list = mutableListOf<ActivityLogListItem>()
+        if (displayProgress) {
+            list.add(ActivityLogListItem.Header(NOW))
+            if (progressWithDate) {
+                list.add(ActivityLogListItem.Progress(RESTORING_CURRENTLY, RESTORING_DATE_TIME))
+            } else {
+                list.add(ActivityLogListItem.Progress(RESTORING_CURRENTLY, RESTORING_NO_DATE))
+            }
+        }
+        if (!emptyList) {
+            firstItem(rewindDisabled).let {
+                list.add(ActivityLogListItem.Header(it.formattedDate))
+                list.add(it)
+            }
+            list.add(secondItem(rewindDisabled))
+            thirdItem(rewindDisabled).let {
+                list.add(ActivityLogListItem.Header(it.formattedDate))
+                list.add(it)
+            }
+        }
+        if (isLastPageAndFreeSite) {
+            list.add(ActivityLogListItem.Footer)
+        }
+        if (canLoadMore) {
+            list.add(ActivityLogListItem.Loading)
+        }
+
+        if (withFooter) {
+            list.add(ActivityLogListItem.Footer)
+        }
+        return list
+    }
+
+    private fun firstItem(rewindDisabled: Boolean) = ActivityLogListItem.Event(
+            model = activityList[0],
+            rewindDisabled = rewindDisabled,
+            backupDownloadFeatureEnabled = false,
+            restoreFeatureEnabled = false
+    )
+
+    private fun secondItem(rewindDisabled: Boolean) = ActivityLogListItem.Event(
+            model = activityList[1],
+            rewindDisabled = rewindDisabled,
+            backupDownloadFeatureEnabled = false,
+            restoreFeatureEnabled = false
+    )
+
+    private fun thirdItem(rewindDisabled: Boolean) = ActivityLogListItem.Event(
+            model = activityList[2],
+            rewindDisabled = rewindDisabled,
+            backupDownloadFeatureEnabled = false,
+            restoreFeatureEnabled = false
+    )
+
     private suspend fun assertFetchEvents(canLoadMore: Boolean = false) {
         verify(store).fetchActivities(fetchActivityLogCaptor.capture())
 
         fetchActivityLogCaptor.lastValue.apply {
-            assertEquals(this.loadMore, canLoadMore)
-            assertEquals(this.site, site)
+            assertEquals(canLoadMore, loadMore)
+            assertEquals(this@ActivityLogViewModelTest.site, site)
         }
     }
 
-    private fun initializeActivityList(): List<ActivityLogModel> {
-        val birthday = Calendar.getInstance()
-        birthday.set(1985, 8, 27)
-
-        val list = mutableListOf<ActivityLogModel>()
-        val activity = ActivityLogModel(
-                "", "", null, "", "", "",
-                "", true, "", birthday.time
-        )
-        list.add(activity)
-        list.add(activity.copy(rewindable = false))
+    private fun event() = ActivityLogListItem.Event(
```

Comment:
❓: Functions with no arguments might be interchangeable with read-only properties. Any reason why we used a function `event()` here? Might help me understand when to prefer one over another.

Reply:
Thanks for the question @ashiagr ! And what a great one indeed. 🌟

You will understand the answer quickly if you check the other such function further above, the `activity()` one, which accept parameters, which all of them default to some value. So, in the end you can use this function as `activity()` and get the default behavior of the (data) class, or update say the `rewindable` to `false` and get an activity that is not rewindable. With the `event()` I went with the same approach, but didn't need to add any variables (yet).

To be honest, there is a longer story to that. This is actually a `stub/fake` pattern that can be applied to any codebase, in terms of testing, where ALL those suchlike function can be created in one place, with all its parameters being available from the function, with default, so that whenever a developer needs to create say an `activity` for their test, they would just import this helper function from that common place and right away use it as `activity()`, without the need to do all this boilerplate setup over and over again. This really removes the clutter in setting up tests by a lot. Also, this helps removing the need to use `@Mock` on a `Site` class, which is not recommended, and have a common `site()` test helper function to have a site created as a stub/fake and not as a mock.

Does the above make sense or am I mumbling again? 😅

Reply:
Thank you for an elaborate answer 🙏, it does make sense!